shuttle/src/sync/mod.rs

```diff
@@ -21,7 +22,99 @@ pub use rwlock::RwLock;
 pub use rwlock::RwLockReadGuard;
 pub use rwlock::RwLockWriteGuard;
 
+use std::hash::{Hash, Hasher};
+use std::panic::Location;
 pub use std::sync::{LockResult, PoisonError, TryLockError, TryLockResult};
 
 // TODO implement true support for `Arc`
 pub use std::sync::{Arc, Weak};
+
+const CONST_CONTEXT_SIGNATURE: u64 = 0;
+
+/// A stable signature for identifying resources in Shuttle tests.
+/// This is used internally to track and differentiate between different resource instances
+/// *across* shuttle iterations. This signature hash is not *guaranteed* to be unique (two
+/// different resources may have the same hash). In particular, resources which provide
+/// `const` constructors cannot provide any *runtime* context to their signatures, so any
+/// two resources created at the same source location will collide.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub(crate) enum ResourceSignature {
```

Comment:
Instead of having this enum, how about having instead

`enum ResourceSignatureType {
    Atomic,
    BatchSemaphore,
    Barrier,
    ...
}`

and then making `ResourceSignature` a wrapper around `ResourceSignatureType` and `ResourceSignatureData`.

This would also make it possible to use separate counters per resource type when generating the signatures, so when debugging, you would know "this is the 3rd RwLock created by this task"